ui/org.eclipse.pde.launching/src/org/eclipse/pde/internal/launching/launcher/LaunchConfigurationHelper.java

```diff
@@ -229,40 +233,28 @@ private static String computeOSGiBundles(String bundleList, Map<String, List<IPl
 		if (bundles.get(IPDEBuildConstants.BUNDLE_SIMPLE_CONFIGURATOR) != null)
 			return "org.eclipse.equinox.simpleconfigurator@1:start"; //$NON-NLS-1$
 
-		StringBuilder buffer = new StringBuilder();
-		Set<String> initialBundleSet = new HashSet<>();
-		StringTokenizer tokenizer = new StringTokenizer(bundleList, ","); //$NON-NLS-1$
+		Map<IPluginModelBase, String> allBundles = new HashMap<>(bundlesWithStartLevels);
+		StringTokenizer tokenizer = new StringTokenizer(initialBundleList, ","); //$NON-NLS-1$
 		while (tokenizer.hasMoreTokens()) {
 			String token = tokenizer.nextToken();
 			int index = token.indexOf('@');
 			String id = index != -1 ? token.substring(0, index) : token;
-			if (bundles.containsKey(id)) {
-				if (buffer.length() > 0)
-					buffer.append(',');
-				buffer.append(id);
-				if (index != -1 && index < token.length() - 1)
-					buffer.append(token.substring(index));
-				initialBundleSet.add(id);
+			IPluginModelBase model = getLatestModel(id, bundles);
+			if (model != null) {
+				String startInfo = index + 1 < token.length() ? token.substring(index + 1) : ""; //$NON-NLS-1$
+				allBundles.put(model, startInfo);
```

Comment:
```suggestion
				allBundles.putIfAbsent(model, startInfo);
```
Otherwise the values specified in the launch config are always overwritten by the values in the given `initialBundleList`.
But I think that this is actually not sufficient because existing entries often have `default:default` as value. So the value should be merged with an existing value where only `default` for the start-level/autostart property is overwritten and a non-default value is left unchanged.
Map.merge with a corresponding remapping function could do that job.